contrib/babelfishpg_tsql/src/dbcmds.c

```diff
@@ -868,3 +904,169 @@ get_owner_of_db(const char *dbname)
+	bbf_set_current_user("sysadmin");
+
+	old_dbid = get_cur_db_id();
+	old_dbname = get_cur_db_name();
+	set_cur_db(dbid, dbname);
+
+	PG_TRY();
+	{
+		PlannedStmt *wrapper;
+		Node *stmt = ((RawStmt *) linitial(parsetree_list))->stmt;
+		update_CreateSchemaStmt(stmt, phys_schema_name, phys_role);
+
+		wrapper = makeNode(PlannedStmt);
+		wrapper->commandType = CMD_UTILITY;
+		wrapper->canSetTag = false;
+		wrapper->utilityStmt = stmt;
+		wrapper->stmt_location = 0;
+		wrapper->stmt_len = 0;
+
+		ProcessUtility(wrapper,
+					query.data,
+					false,
+					PROCESS_UTILITY_SUBCOMMAND,
+					NULL,
+					NULL,
+					None_Receiver,
+					NULL);
+
+		/* make sure later steps can see the object created here */
+		CommandCounterIncrement();
+	}
+	PG_CATCH();
+	{
+		bbf_set_current_user(prev_current_user);
+		set_cur_db(old_dbid, old_dbname);
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	bbf_set_current_user(prev_current_user);
+	set_cur_db(old_dbid, old_dbname);
+
+}
+
+/*
+* This function is only being used for the purpose of the upgrade script to add
+* the guest schema for each database if the database does not have the guest schema yet.
+*/
+PG_FUNCTION_INFO_V1(create_guest_schema_for_all_dbs);
+Datum create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+{
+	Relation		sysdatabase_rel;
+	TableScanDesc 	scan;
+	HeapTuple 		tuple;
+	const char		*sql_dialect_value_old;
+	const char		*tsql_dialect = "tsql";
+	Form_sysdatabases   bbf_db;
+	const char		*dbname;
+	bool creating_extension_backup = creating_extension;
+
+	sql_dialect_value_old = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
+
+	PG_TRY();
+	{
+		set_config_option("babelfishpg_tsql.sql_dialect", tsql_dialect,
+							(superuser() ? PGC_SUSET : PGC_USERSET),
+							PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+
+		/*
+		* Since this is part of upgrade script, PG assumes we would like to set
+		* the babelfish extension depend on this new schema. This is not true
+		* so we tell PG not to set any dependency for us.
+		* Check recordDependencyOnCurrentExtension() for more information.
+		*/
+		creating_extension = false;
+
+		sysdatabase_rel = table_open(sysdatabases_oid, RowExclusiveLock);
+		scan = table_beginscan_catalog(sysdatabase_rel, 0, NULL);
+		tuple = heap_getnext(scan, ForwardScanDirection);
+
+		while (HeapTupleIsValid(tuple)) 
+		{
+			bbf_db = (Form_sysdatabases) GETSTRUCT(tuple);
+			dbname = text_to_cstring(&(bbf_db->name));
+
+			create_schema_if_not_exists(bbf_db->dbid, dbname, "guest", "guest");
+
+			tuple = heap_getnext(scan, ForwardScanDirection);
+		}
+		table_endscan(scan);
+		table_close(sysdatabase_rel, RowExclusiveLock);
+
+		creating_extension = creating_extension_backup;
+		set_config_option("babelfishpg_tsql.sql_dialect", sql_dialect_value_old,
+							(superuser() ? PGC_SUSET : PGC_USERSET),
+							PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+	}
+	PG_CATCH();
+	{
+		creating_extension = creating_extension_backup;
```

Comment:
in cast of exception, need to end scan , close table as well

Reply:
PG should be able to release the locks and resources with its error handling. Also adding in the endscan and close in PG_CATCH is not safe. Consider the case where for some unknown reason it crashed on line 1056, then PG_CATCH would do a double endscan and close which would cause the segfault


```052         table_endscan(scan);
1053         table_close(sysdatabase_rel, RowExclusiveLock);
1054
1055         creating_extension = creating_extension_backup;
1056         set_config_option("babelfishpg_tsql.sql_dialect", sql_dialect_value_old,
1057                             (superuser() ? PGC_SUSET : PGC_USERSET),
1058                             PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
1059     }
1060     PG_CATCH();
1061     {
1062         creating_extension = creating_extension_backup;
1063         set_config_option("babelfishpg_tsql.sql_dialect", sql_dialect_value_old,
1064                             (superuser() ? PGC_SUSET : PGC_USERSET),
1065                             PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
1066
1067         PG_RE_THROW();
1068     }```

Reply:
similar logic in `drop_all_dbs()`